// @pixi/ui/lib/DoubleSlider.js
'use strict';

var SliderBase = require('./SliderBase.js');
var typedSignals = require('typed-signals');

class DoubleSlider extends SliderBase.SliderBase {
  constructor(options) {
    super(options);
    /** Signal that fires when value have changed. */
    this.onChange = new typedSignals.Signal();
    /** Signal that fires when value is changing. */
    this.onUpdate = new typedSignals.Signal();
    this.options = options;
    this.setInitialState();
  }
  setInitialState() {
    this.validateValues();
    const { value1, value2 } = this.options;
    this.updateProgress(value1, value2);
    this.value2 = value2;
    this.value1 = value1;
  }
  updateProgress(value1 = this.value1, value2 = this.value2) {
    this.progressStart = (value1 - this.min) / (this.max - this.min) * 100;
    this.progress = (value2 - this.min) / (this.max - this.min) * 100;
  }
  validateValues() {
    if (!this.options.value1) {
      this.options.value1 = this.min;
    }
    if (!this.options.value2) {
      this.options.value2 = this.options.max;
    }
    if (this.options.value2 < this.options.value1) {
      this.options.value2 = this.options.value1;
    }
    if (this.options.value1 < this.options.min) {
      this.options.value1 = this.options.min;
    }
    if (this.options.value1 > this.options.max) {
      this.options.value1 = this.options.max;
    }
    if (this.options.value2 > this.options.max) {
      this.options.value2 = this.options.max;
    }
  }
  /** Returns left value. */
  get value1() {
    return this._value1;
  }
  /** Sets left value. */
  set value1(value1) {
    if (value1 === this._value1)
      return;
    if (value1 < this.min)
      value1 = this.min;
    if (value1 > this._value2)
      value1 = this._value2;
    this._value1 = value1;
    this.updateSlider1();
    this.onUpdate?.emit(this.value1, this.value2);
  }
  /** Returns right value. */
  get value2() {
    return this._value2;
  }
  /** Sets right value. */
  set value2(value2) {
    if (value2 === this._value2)
      return;
    if (value2 < this._value1)
      value2 = this._value1;
    if (value2 > this.max)
      value2 = this.max;
    this._value2 = value2;
    this.updateSlider2();
    this.onUpdate?.emit(this.value1, this.value2);
  }
  update(event) {
    if (!this.dragging)
      return;
    const obj = event.currentTarget;
    const { x } = obj.parent.worldTransform.applyInverse(event.global);
    const slider1Dist = Math.abs(x - this._slider1.x - this._slider1.width);
    const slider2Dist = Math.abs(x - this._slider2.x);
    if (!this.activeValue) {
      if (this.slider1 && x < this.slider1.x) {
        this.activeValue = "value1";
      } else if (this.slider2 && x > this.slider2.x) {
        this.activeValue = "value2";
      } else {
        this.activeValue = slider1Dist < slider2Dist ? "value1" : "value2";
      }
    }
    const progress = this.validate(x / this.bg.width * 100);
    if (this.activeValue === "value1") {
      this.progressStart = progress;
      this.value1 = this.min + (this.max - this.min) / 100 * progress;
      this.updateProgress(this.value1, this.value2);
    } else {
      this.progress = progress;
      this.value2 = this.min + (this.max - this.min) / 100 * progress;
    }
  }
  endUpdate() {
    super.endUpdate();
    this.activeValue = null;
  }
  change() {
    this.onChange?.emit(this.value1, this.value2);
  }
  /**
   * Set Slider1 instance.
   * @param value - Container or string with texture name.
   */
  set slider1(value) {
    super.slider1 = value;
    this.updateSlider1();
  }
  /** Get Slider1 instance. */
  get slider1() {
    return this._slider1;
  }
  /**
   * Sets Slider instance.
   * @param value - Container or string with texture name.
   */
  set slider2(value) {
    super.slider2 = value;
    this.updateSlider2();
  }
  /** Get Slider2 instance. */
  get slider2() {
    return this._slider2;
  }
  updateSlider1() {
    this._slider1.x = (this.bg.width - this._slider1.width) / 100 * this.progressStart;
    if (this._slider2 && this._slider1.x > this._slider2.x) {
      this._slider1.x = this._slider2.x;
    }
    if (this.options?.showValue) {
      this.value1Text.text = `${Math.round(this.value1)}`;
      const sliderPosX = this._slider1.x + this._slider1.width / 2;
      const sliderPosY = this._slider1.y;
      this.value1Text.x = sliderPosX + (this.options.valueTextOffset?.x ?? 0);
      this.value1Text.y = sliderPosY + (this.options.valueTextOffset?.y ?? 0);
    }
  }
  updateSlider2() {
    this._slider2.x = (this.bg.width - this._slider2.width) / 100 * this.progress;
    if (this._slider2.x < this._slider1.x) {
      this._slider2.x = this._slider1.x;
    }
    if (this.options?.showValue) {
      this.value2Text.text = `${Math.round(this.value2)}`;
      const sliderPosX = this._slider2.x + this._slider2.width / 2;
      const sliderPosY = this._slider2.y;
      this.value2Text.x = sliderPosX + (this.options.valueTextOffset?.x ?? 0);
      this.value2Text.y = sliderPosY + (this.options.valueTextOffset?.y ?? 0);
    }
  }
}

exports.DoubleSlider = DoubleSlider;
