// @pixi/ui/lib/Select.js
'use strict';

var display = require('@pixi/display');
var graphics = require('@pixi/graphics');
var text = require('@pixi/text');
var typedSignals = require('typed-signals');
var FancyButton = require('./FancyButton.js');
var ScrollBox = require('./ScrollBox.js');
var view = require('./utils/helpers/view.js');

const defaultVisibleItems = 5;
class Select extends display.Container {
  constructor(options) {
    super();
    this.onSelect = new typedSignals.Signal();
    if (options) {
      this.init(options);
    }
  }
  /**
   * Initiates Select.
   * @param root0
   * @param root0.closedBG
   * @param root0.textStyle
   * @param root0.items
   * @param root0.openBG
   * @param root0.selected
   * @param root0.selectedTextOffset
   * @param root0.scrollBox
   * @param root0.visibleItems
   */
  init({ closedBG, textStyle, items, openBG, selected, selectedTextOffset, scrollBox, visibleItems }) {
    if (this.openView && this.openView !== openBG) {
      this.removeChild(this.openView);
    }
    if (!this.openButton) {
      this.openButton = new FancyButton.FancyButton({
        defaultView: view.getView(closedBG),
        text: new text.Text(items?.items ? items.items[0] : "", textStyle),
        textOffset: selectedTextOffset
      });
      this.openButton.onPress.connect(() => this.toggle());
      this.addChild(this.openButton);
    } else {
      this.openButton.views = {
        defaultView: view.getView(closedBG),
        text: new text.Text(items?.items ? items.items[0] : "", textStyle)
      };
      this.openButton.textOffset = selectedTextOffset;
    }
    if (this.openView !== openBG) {
      this.openView = view.getView(openBG);
      this.openView.visible = false;
      this.addChild(this.openView);
    }
    if (!this.closeButton) {
      this.closeButton = new FancyButton.FancyButton({
        defaultView: new graphics.Graphics().beginFill(0, 1e-5).drawRect(0, 0, this.openButton.width, this.openButton.height),
        text: new text.Text(items?.items ? items.items[0] : "", textStyle),
        textOffset: selectedTextOffset
      });
      this.closeButton.onPress.connect(() => this.toggle());
      this.openView.addChild(this.closeButton);
    } else {
      this.closeButton.views = {
        defaultView: new graphics.Graphics().beginFill(0, 1e-5).drawRect(0, 0, this.openButton.width, this.openButton.height),
        text: new text.Text(items?.items ? items.items[0] : "", textStyle)
      };
      this.openButton.textOffset = selectedTextOffset;
    }
    if (!this.scrollBox) {
      this.scrollBox = new ScrollBox.ScrollBox();
      this.openView.addChild(this.scrollBox);
    } else {
      this.scrollBox.removeItems();
    }
    this.scrollBox.init({
      type: "vertical",
      elementsMargin: 0,
      width: this.openButton.width,
      height: this.openButton.height * (visibleItems ?? defaultVisibleItems),
      radius: 0,
      padding: 0,
      ...scrollBox
    });
    this.scrollBox.y = this.openButton.height;
    if (scrollBox?.offset) {
      this.scrollBox.x = scrollBox.offset.x ?? 0;
      this.scrollBox.y += scrollBox.offset.y ?? 0;
    }
    this.convertItemsToButtons(items).forEach((button, id) => {
      const text = button.text;
      if (id === selected) {
        this.openButton.text = text;
        this.closeButton.text = text;
      }
      button.onPress.connect(() => {
        this.value = id;
        this.onSelect.emit(id, text);
        this.openButton.text = text;
        this.closeButton.text = text;
        this.close();
      });
      this.scrollBox.addItem(button);
    });
  }
  /** Toggle the select state (open if closed, closes - id open). */
  toggle() {
    this.openView.visible = !this.openView.visible;
    this.openButton.visible = !this.openButton.visible;
  }
  /** Show dropdown. */
  open() {
    this.openView.visible = true;
    this.openButton.visible = false;
  }
  /** Hide dropdown. */
  close() {
    this.openView.visible = false;
    this.openButton.visible = true;
  }
  convertItemsToButtons({
    items,
    backgroundColor,
    hoverColor,
    width,
    height,
    textStyle,
    radius
  }) {
    const buttons = [];
    items.forEach((item) => {
      const defaultView = new graphics.Graphics().beginFill(backgroundColor).drawRoundedRect(0, 0, width, height, radius);
      const color = hoverColor ?? backgroundColor;
      const hoverView = new graphics.Graphics().beginFill(color).drawRoundedRect(0, 0, width, height, radius);
      const text$1 = new text.Text(item, textStyle);
      const button = new FancyButton.FancyButton({ defaultView, hoverView, text: text$1 });
      buttons.push(button);
    });
    return buttons;
  }
}

exports.Select = Select;
