// @pixi/ui/lib/SliderBase.js
'use strict';

var display = require('@pixi/display');
var sprite = require('@pixi/sprite');
var text = require('@pixi/text');
var ProgressBar = require('./ProgressBar.js');
var view = require('./utils/helpers/view.js');

class SliderBase extends ProgressBar.ProgressBar {
  constructor(options) {
    super();
    this.dragging = 0;
    /** Minimal value. */
    this.min = 0;
    /** Maximal value. */
    this.max = 100;
    this.setBackground(options.bg);
    if (options.fill) {
      this.setFill(options.fill, options.fillOffset);
    }
    this.settings = options;
    this.slider1 = options.slider1;
    this.slider2 = options.slider2;
    this.min = options.min ?? 0;
    this.max = options.max ?? 100;
    this.activate();
  }
  /**
   * Sets Slider1 instance.
   * @param value - Container or string with texture name.
   */
  set slider1(value) {
    if (!value)
      return;
    if (this._slider1) {
      this.slider1.removeAllListeners();
      this.removeChild(this._slider1);
      this.slider1.destroy();
    }
    this._slider1 = this.createSlider(value);
    this._slider1.eventMode = "static";
    this._slider1.on("pointerdown", this.startUpdate, this).on("globalpointermove", this.update, this).on("pointerup", this.endUpdate, this).on("pointerupoutside", this.endUpdate, this);
    if (this.settings.showValue && !this.value1Text) {
      this.value1Text = new text.Text("", this.settings.valueTextStyle || { fill: 16777215 });
      this.value1Text.anchor.set(0.5);
      this.addChild(this.value1Text);
    }
  }
  /** Get Slider1 instance. */
  get slider1() {
    return this._slider1;
  }
  /**
   * Sets Slider2 instance.
   * @param value - Container or string with texture name.
   */
  set slider2(value) {
    if (!value)
      return;
    if (this._slider2) {
      this.slider2.removeAllListeners();
      this.removeChild(this._slider2);
      this.slider2.destroy();
    }
    this._slider2 = this.createSlider(value);
    this._slider2.eventMode = "static";
    this._slider2.on("pointerdown", this.startUpdate, this).on("globalpointermove", this.update, this).on("pointerup", this.endUpdate, this).on("pointerupoutside", this.endUpdate, this);
    if (this.settings.showValue && !this.value2Text) {
      this.value2Text = new text.Text("", this.settings.valueTextStyle || { fill: 16777215 });
      this.value2Text.anchor.set(0.5);
      this.addChild(this.value2Text);
    }
  }
  /** Get Slider2 instance. */
  get slider2() {
    return this._slider2;
  }
  /**
   * Set bg.
   * @param bg
   */
  setBackground(bg) {
    if (this.bg) {
      this.bg.removeAllListeners();
    }
    super.setBackground(bg);
    this.activate();
  }
  activate() {
    this.bg.eventMode = "static";
    this.bg.on("pointerdown", this.startUpdate, this).on("globalpointermove", this.update, this).on("pointerup", this.endUpdate, this).on("pointerupoutside", this.endUpdate, this);
    if (this.fill) {
      this.fill.eventMode = "none";
    }
    if (this.value1Text) {
      this.value1Text.eventMode = "none";
    }
    if (this.value2Text) {
      this.value2Text.eventMode = "none";
    }
  }
  createSlider(sliderData) {
    const slider = view.getView(sliderData);
    slider.x = slider.width / 2;
    const container = new display.Container();
    container.addChild(slider);
    if (slider instanceof sprite.Sprite) {
      slider.anchor.set(0.5);
    }
    container.y = this.bg.height / 2;
    this.addChild(container);
    return container;
  }
  startUpdate(event) {
    this.dragging = 1;
    this.startUpdateValue1 = this._value1;
    this.startUpdateValue2 = this._value2;
    this.update(event);
  }
  endUpdate() {
    if (!this.dragging)
      return;
    this.dragging = 0;
    if (this.startUpdateValue1 !== this._value1 || this.startUpdateValue2 !== this._value2) {
      this.change();
    }
    this.startUpdateValue1 = null;
    this.startUpdateValue2 = null;
  }
  /* Called when dragging started and on every move. */
  update(_event) {
  }
  /** Called when dragging stopped. */
  change() {
  }
}

exports.SliderBase = SliderBase;
